### ISSUE TITLE
Faster LHS checking - lattice problems

### ISSUE THREAD

**Issue by Timeroot** [ShaharGottlieb/MasseyRamanujan#12]
Hi, I think you could speed up your LHS checking through lattice algorithms. This isn't a novel idea of mine, but seems applicable here. After realizing it could be applied here, I googled it and found other people talking about the same trick, e.g. https://wstein.org/books/ant/ant/node15.html

The idea is that you have some mysterious real number from your RHS gcf -- say, 2.2959862202 -- and I want to look for a rational expression (that includes other irrational quantities, of course) that equals this. Call the real number x. I'm looking for any equation of the form (say),

(p1 + p2 * e + p3 * pi) / (q1 + q2 * e + q3 * pi) = x.

Where p's and q's are the unknowns that I'm searching for. Multiply through by the denominator, and subtract the RHS from both sides:

p1 + p2 * e + p3 * pi + q1 * (-x) + q2 * (-e*x) + q3 * (-pi*x)  =  0

so that it's now a problem about finding a linear dependence over Q (or equivalently, over Z) of the numbers {1, e, pi, x, e*x, pi*x}. The reals can be viewed as a 1D vector space, and we're looking for a "very short" vector. To turn this into a workable algorithm, you want your lattice to have at least the same dimension in space as the number of basis vectors, so you add one dimension for each quantity. Then you scale up the "error" coordinate to penalize it a lot. Here, I will multiply by some larger number M. The new vectors are:

```
v1=(M*1, 1,0,0,0,0,0)
v2=(M*e, 0,1,0,0,0,0)
v3=(M*pi, 0,0,1,0,0,0)
v4=(M*x, 0,0,0,1,0,0)
v5=(M*e*x, 0,0,0,0,1,0)
v6=(M*pi*x, 0,0,0,0,0,1)
```

This set of 6 vectors generates a subspace of R^7. Now set, say, M=10000, and you look for the shortest vector. A short vector will necessarily have a very small "error coordinate". If you use x=2.2959862202, you'll find for instance that the vector

V = 2*v1 + v2 + v3 - 3 v4 + v5 - v6

is very short! Even though all the input vectors (v1 ... v6) had length at least 10000, the vector V is just

(0.000000000001, 2, 1, 1, -3, 1, -1)

and so has a length of just 4.12. Reading off the coefficients of V, we find that x = (2+e+pi)/(3-e+pi).

Thankfully, there are standard algorithms such as LLL which enable this kind of shortest vector discovery. They're not guaranteed to find short vectors, but in a large fraction of cases they do very well. In the case where they fail to find a solution, randomly permuting the vectors and trying again several times (say, 10 times) will often help. For instance, here is Mathematica code to solve this problem:

```
x = 2.295986220225457575964614984963958313;
M = 1000000;
v1 = {M*1, 1, 0, 0, 0, 0, 0};
v2 = {M*E, 0, 1, 0, 0, 0, 0};
v3 = {M*Pi, 0, 0, 1, 0, 0, 0};
v4 = {M*x, 0, 0, 0, 1, 0, 0};
v5 = {M*E*x, 0, 0, 0, 0, 1, 0};
v6 = {M*Pi*x, 0, 0, 0, 0, 0, 1};

LatticeReduce[Round[{v1, v2, v3, v4, v5, v6}]][[1]]
```

The "Round" here is because Mathematica's LatticeReduce algorithm only accepts integers, so I use a large "M" value and then round the irrational numbers (like M*E, M*x) off to their nearest integers. This spits out "{-2, -2, -1, -1, 3, -1, 1}" instantly, which gives the solution.

As noted also in that link above, this technique can be extended to finding not just rational functions of x, but also algebraic equations involving x. For instance, you're very unlikely to recognize the number 1.35048396195 on its own, but it turns out to be the solution to the quintic x^5 - x = pi. If you search using the "vectors" {1, Pi, x, x^2, x^3, x^4, x^5}, you'll quickly find this solution. There are known continued fractions to produce e.g. any rational root of a rational number, I believe there are also continued fractions for quantities like the roots of x^3 - x - 1. So I would expect this technique to be somewhat fruitful too.

I believe with high confidence that this is the same kind of algorithm that Wolfram|Alpha uses to find its "closed form approximations" to numbers, and indeed dropping 2.29598622022 it finds the rational function of E and Pi quickly, and it regularly suggests algebraic roots as well.

This would be a welcome alternative to hashmap-based LHS searching! :) Please contact me if you want to discuss this more.

**Comment by ShaharGottlieb** [comment]
Hi Alex, thanks for reaching out!

As you mentioned, this kind of idea could be very useful and I will look up
the link you've sent.
We even tried "playing" with these kinds of ideas. some code even remained
in our repo - unused (
https://github.com/ShaharGottlieb/MasseyRamanujan/blob/master/ramanujan/utils/mobius.py,
line 281).
Is this the kind of magic you are talking about?

The problem with this is that the complexity of finding integer relations
is much more heavier than looking up a hash table.
If we search N possible RHS then we need to apply a Lattice Reduction
algorithm N times (which is impractical).

BUT - good news, we are looking into some way to reduce N and find only
"interesting" RHS candidates (that have some mathematical property we call
"factorial reduction").
This kind of direction has a lot to do with "pure mathematics" as we need
to discard most of the RHS candidates before.
In the future we definitely plan to use more sophisticated ways to avoid
saving a huge hash table.

Best regards and thanks again,

Shahar

**Comment by Timeroot** [comment]
Yes, something like that! But I think LLL -- or, as @kaminer pointed out to me in a separate email, PSLQ -- can find a broader class of solutions, more efficiently than that integer linear programming instance. I'm playing around with it a bit. On small numbers of constants (say, 6) to search through, PSLQ seems very fast, like a few seconds. The time scales up pretty quickly with the number of constants though, taking a minute with 10 constants, and tens of minutes with 15 constants. (These time estimates are also a bit complicated because, as you decrease the number of constants, all calculations can be done with less working precision to avoid spurious results, which reduces the runtime further.)

I'm trying to tune my code a bit to see what I can get to work, and will submit a PR when I have something, but I think a promising approach would be to run PSLQ on several different small sets of constants. For instance, checking for quadratic polynomials in each of 20 different constants separately (which would include all of the "mobius" values), and for rational functions that involve any two of the 10 more common constants.
